src/goto-programs/goto_convert.cpp

```diff
@@ -857,25 +857,40 @@ void goto_convertt::convert_assume(
   dest.add(goto_programt::make_assumption(op, code.source_location()));
 }
 
-void goto_convertt::convert_loop_invariant(
+void goto_convertt::convert_loop_invariant_and_decreases(
   const codet &code,
   goto_programt::targett loop,
   const irep_idt &mode)
 {
   exprt invariant=
     static_cast<const exprt&>(code.find(ID_C_spec_loop_invariant));
+  exprt variant = static_cast<const exprt&>(code.find(ID_C_spec_decreases));
 
-  if(invariant.is_nil())
-    return;
-
-  if(has_subexpr(invariant, ID_side_effect))
+  if(!invariant.is_nil())
   {
-    throw incorrect_goto_program_exceptiont(
-      "Loop invariant is not side-effect free.", code.find_source_location());
+    if(has_subexpr(invariant, ID_side_effect))
+    {
+      throw incorrect_goto_program_exceptiont(
+        "Loop invariant is not side-effect free.", code.find_source_location());
+    }
+
+    PRECONDITION(loop->is_goto());
+    loop->guard.add(ID_C_spec_loop_invariant).swap(invariant);
   }
 
-  PRECONDITION(loop->is_goto());
-  loop->guard.add(ID_C_spec_loop_invariant).swap(invariant);
+  if(!variant.is_nil())
+  {
+    // Interestingly enough, a function call is considered a side effect even if
+    // the callee entails no side effects. 
```

Comment:
I'd remove these two  lines -- this is a known issue. If you want to document this, let's document that before the first call to `has_subexpr` as well. May be add a `TODO` at line 868 and explain this, if you want to.

Reply:
OK. I will just remove these two lines of comments.